ownCloud/Client/Actions/Actions+Extensions/UploadPhotosAction.swift

```diff
@@ -75,16 +79,52 @@ class UploadPhotosAction: UploadBaseAction {
 		}
 	}
 
-	func presentImageGalleryPicker() {
+	private func presentImageGalleryPicker() {
 		OnMainThread {
 			if let viewController = self.context.viewController {
 				let photoAlbumViewController = PhotoAlbumTableViewController()
 				photoAlbumViewController.selectionCallback = { (assets) in
-					for asset in assets {
-						self.upload(asset: asset)
-					}
 
 					self.completed()
+
+					let queue = DispatchQueue.global(qos: .userInitiated)
+
+					queue.async {
+
+						self.core?.perform(inRunningCore: { (runningCoreCompletion) in
+
+							let uploadGroup = DispatchGroup()
+							var uploadFailed = false
+
+							for asset in assets {
+								if uploadFailed == false {
+									// Upload image on a background queue
+									uploadGroup.enter()
+
+									self.upload(asset: asset, completion: { (success, uploadFinished) in
+										if !success {
+											uploadFailed = true
+										}
+										if uploadFinished {
+											uploadGroup.leave()
+										}
+									})
+
+									// Avoid submitting to many jobs simultaneously to reduce memory pressure
+									_ = uploadGroup.wait(timeout: .now() + 0.5)
+
+								} else {
+									// Escape on first failed download
+									break
+								}
+							}
+
+							uploadGroup.notify(queue: queue, execute: {
+								runningCoreCompletion()
```

Comment:
(3/3) … so that `runningCoreCompletion()` is never called, either in that case, keeping the core running indefinitely.

Regarding general use of this API, it should be as brief as possible.

Since it's the job of the Core/Sync Engine to ensure the files are uploaded once submitted to it, it's safe to call `runningCoreCompletion()` when all `placeholderCompletionHandler`s were called.

Otherwise, the core is kept from stopping until all uploads have finished, eventually blocking users that have "logged out" from the server from logging back in.

Reply:
@felix-schwarz implemented as suggested